src/censored.jl

```diff
@@ -0,0 +1,278 @@
+    return censored(d, promote(l, u)...)
+end
+censored(d::UnivariateDistribution, l::T, u::T) where {T <: Real} = Censored(d, l, u)
+
+"""
+    Censored
+
+Generic wrapper for a censored distribution.
+"""
+struct Censored{D<:UnivariateDistribution, S<:ValueSupport, T <: Real} <: UnivariateDistribution{S}
+    uncensored::D      # the original distribution (uncensored)
+    lower::T      # lower bound
+    upper::T      # upper bound
+    function Censored(d::UnivariateDistribution, l::T, u::T) where {T <: Real}
+        l ≤ u || error("the lower bound must be less than or equal to the upper bound")
+        new{typeof(d), value_support(typeof(d)), T}(d, l, u)
+    end
+end
+Censored{D,S,T}(l::T, u::T, params...) where {D,S,T} = Censored(D(params...), l, u)
+
+function censored(d::Censored, l::T, u::T) where {T <: Real}
+    return Censored(d.uncensored, max(l, d.lower), min(u, d.upper))
+end
+
+params(d::Censored) = (d.lower, d.upper, params(d.uncensored)...)
+partype(d::Censored) = Base.promote_eltype(partype(d.uncensored), d.lower)
+Base.eltype(::Type{Censored{D,S,T}}) where {D,S,T} = Base.promote_eltype(eltype(D), T)
+
+#### Range and Support
+
+islowerbounded(d::Censored) = islowerbounded(d.uncensored) || isfinite(d.lower)
+isupperbounded(d::Censored) = isupperbounded(d.uncensored) || isfinite(d.upper)
+
+minimum(d::Censored) = max(minimum(d.uncensored), d.lower)
+maximum(d::Censored) = min(maximum(d.uncensored), d.upper)
+
+function insupport(d::Censored{<:UnivariateDistribution}, x::Real)
+    l = d.lower
+    u = d.upper
+    return x == l || x == u || (l < x < u && insupport(d.uncensored, x))
+end
+
+#### Show
+
+function show(io::IO, d::Censored)
+    print(io, "Censored(")
+    d0 = d.uncensored
+    uml, namevals = _use_multline_show(d0)
+    uml ? show_multline(io, d0, namevals; newline=false) : show_oneline(io, d0, namevals)
+    print(io, ", range=(", d.lower, ", ", d.upper, "))")
+end
+
+_use_multline_show(d::Censored) = _use_multline_show(d.uncensored)
+
+
+#### Statistics
+
+quantile(d::Censored, p::Real) = clamp(quantile(d.uncensored, p), d.lower, d.upper)
+
+median(d::Censored) = clamp(median(d.uncensored), d.lower, d.upper)
+
+function mean(d::Censored)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    dtrunc = truncated(d.uncensored, lower, upper)
+    prob_lower = if value_support(typeof(d)) === Discrete && isfinite(lower)
+        cdf(d0, lower) - pdf(d0, lower)
+    else
+        cdf(d0, lower)
+    end
+    prob_upper = ccdf(d0, upper)
+    prob_interval = 1 - (prob_lower + prob_upper)
+
+    μ = prob_interval * mean(dtrunc)
+    if isfinite(lower)
+        μ += prob_lower * lower
+    end
+    if isfinite(upper)
+        μ += prob_upper * upper
+    end
+    return μ
+end
+
+function var(d::Censored)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    dtrunc = truncated(d.uncensored, lower, upper)
+    prob_lower = if value_support(typeof(d)) === Discrete && isfinite(lower)
```

Comment:
AFAIK it should be possible to evaluate the pdf and cdf of distributions also at +-Inf (e.g., `pdf(Poisson(), -Inf) == 0`), so the check for finite values could be omitted.